.github/workflows/checker.yml

```diff
@@ -0,0 +1,114 @@
+name: Checker Script CI
+
+on:
+  pull_request:
+    branches:
+      - main
+      - master
+
+jobs:
+  run-checker:
+    runs-on: ubuntu-latest
+
+    steps:
+      # Step 1: Checkout the code from the repository
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      # Step 2: Cache cargo dependencies to avoid reinstalling every time
+      - name: Cache cargo registry
+        uses: actions/cache@v3
+        with:
+          path: |
+            ~/.cargo/registry
+            ~/.cargo/git
+          key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-cargo-
+
+      # Step 3: Cache rust toolchain and binaries (including cargo-stylus)
+      - name: Cache rust toolchain
+        uses: actions/cache@v3
+        with:
+          path: ~/.rustup
+          key: ${{ runner.os }}-rustup-${{ hashFiles('**/rust-toolchain.toml') }}
+          restore-keys: |
+            ${{ runner.os }}-rustup-
+
+      # Step 4: Ensure only rustup manages Rust
+      - name: Ensure only rustup manages Rust
+        run: |
+          sudo apt-get remove --purge cargo rustc || true
+          brew uninstall --force rust || true
+          sudo rm -rf /usr/local/bin/rustc /usr/local/bin/cargo || true
+
+      # Step 5: Install rustup if not present
+      - name: Install rustup
+        run: |
+          curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
+          source $HOME/.cargo/env
+
+      # Step 6: Install Rust toolchain and wasm32 target per directory
+      - name: Install Rust toolchain and wasm32 target per directory
+        run: |
+          for dir in example_code/applications/* example_code/basic_examples/*; do
+            if [[ -d "$dir" && -f "$dir/rust-toolchain.toml" ]]; then
+              echo "Installing Rust toolchain for directory: $dir"
+              toolchain_version=$(grep '^channel' "$dir/rust-toolchain.toml" | cut -d '"' -f 2)
+
+              # Install the toolchain specified in rust-toolchain.toml
+              rustup toolchain install "$toolchain_version"
```

Comment:
Do we need to install the toolchain for every folder? afaik this is defined at the os level (not 100% sure though)